File: src/host/acp/attachments.ts
import { createHash } from 'node:crypto';
import { readFile, stat } from 'node:fs/promises';
import { basename } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import type * as acp from '@agentclientprotocol/sdk';
import type { Attachment, Draft } from '@shared/transcript';
import { MAX_IMAGE_BYTES, MAX_TEXT_BYTES, base64Bytes, extOfMime, imageMimeOf } from '@shared/attachments';
import type { AgentDef } from './AgentRegistry';
import { msg } from '../errors';
import { t } from '../i18n';

// Blob names are content-derived so a transcript can reference a blob before its write completes (agent images are
// referenced synchronously while the store write lands async) and identical payloads share one file
export function blobName(ext: string, bytes: Uint8Array): string {
  return `${createHash('sha256').update(bytes).digest('hex').slice(0, 16)}${ext}`;
}

// Where a session parks attachment payloads (TranscriptStore implements it): the name is what the turn keeps and the webview loads via blobBase,
// the absolute path is what the agent gets told when the content is embedded. Names follow blobName()
export interface BlobStore {
  saveBlob(sessionId: string, ext: string, bytes: Uint8Array): Promise<{ name: string; path: string }>;
  readBlob(sessionId: string, name: string): Promise<Uint8Array>;
  // Absolute path of a saved blob, when the store is on this machine's disk (export links, opening in the editor)
  blobPath?(sessionId: string, name: string): string | undefined;
}

export interface PreparedPrompt {
  blocks: acp.ContentBlock[];
  attachments: Attachment[];
  // What could not be done as asked (a draft dropped for size, a blob that failed to write); the prompt itself still goes out
  problems: string[];
}

// What the agent says it can take in a prompt, plus the per-agent override for a capability that lies.
// An absent capability (or no caps at all) means permitted — the historical behaviour
export interface PromptCaps {
  embeddedContext?: boolean;
  image?: boolean;
  // The agent advertises image: false yet accepts image blocks (Grok) — declared per agent as AgentDef.prompt.imagesRegardless
  imagesRegardless?: boolean;
}

// The prompt capabilities to stage against: the advertised set, defaulting to permitted when the agent said nothing,
// overridden by what the registry knows about the agent
export function promptCapsOf(init: acp.InitializeResponse | undefined, def: AgentDef): PromptCaps {
  const p = init?.agentCapabilities?.promptCapabilities;
  return { embeddedContext: p?.embeddedContext ?? true, image: p?.image ?? true, imagesRegardless: def.prompt?.imagesRegardless };
}

// Turns the composer's text + drafts into the wire prompt and the transcript attachments. Text goes first, then one block per draft:
// images inline as base64 (unless the agent says it takes none — the Grok exception is declared per agent, AgentDef.prompt.imagesRegardless),
// dropped text as an embedded resource whose uri is the blob written to disk (plain marked-up text when embeddedContext is unsupported),
// files as resource_link so the agent reads them itself — except image files, which are read here and sent as pixels.
// Never throws for a single draft: a payload the disk refuses still goes over the wire (the preview is lost), an oversized image is dropped with a note
export async function preparePrompt(sessionId: string, text: string, drafts: Draft[], blobs: BlobStore, caps?: PromptCaps): Promise<PreparedPrompt> {
  const out: PreparedPrompt = { blocks: text ? [{ type: 'text', text }] : [], attachments: [], problems: [] };
  const noImages = caps?.image === false && !caps.imagesRegardless;
  const stage = async (ext: string, bytes: Uint8Array, label: string) => {
    try { return await blobs.saveBlob(sessionId, ext, bytes); }
    catch (e) { out.problems.push(t('host.attachStageFailed', { label, error: msg(e) })); return undefined; }
  };
  for (const d of drafts) {
    if (d.kind === 'image') {
      if (noImages) { out.problems.push(t('host.imageUnsupported', { name: d.name ?? t('common.image') })); continue; }
      if (base64Bytes(d.data) > MAX_IMAGE_BYTES) { out.problems.push(t('host.imageTooBig', { name: d.name ?? t('common.image'), mb: MAX_IMAGE_BYTES >> 20 })); continue; }
      const saved = await stage(extOfMime(d.mimeType), Buffer.from(d.data, 'base64'), d.name ?? t('common.image'));
      out.blocks.push({ type: 'image', mimeType: d.mimeType, data: d.data });
      out.attachments.push({ kind: 'image', blob: saved?.name, mimeType: d.mimeType, name: d.name });
    } else if (d.kind === 'text') {
      // The webview refuses oversized drops too, but the ceiling is enforced here as well: the text is embedded into the prompt verbatim
      const bytes = Buffer.from(d.text, 'utf8');
      if (bytes.byteLength > MAX_TEXT_BYTES) { out.problems.push(t('attach.tooBigText', { name: d.name, kb: MAX_TEXT_BYTES >> 10 })); continue; }
      // The embedded resource is addressed by the blob on disk, so an agent that insists on reading a real file finds one
      const saved = await stage('.txt', bytes, d.name);
      if (caps?.embeddedContext === false) {
        out.blocks.push({ type: 'text', text: `[Attachment: ${d.name}]\n${d.text}\n[End of attachment: ${d.name}]` });
      } else {
        const uri = saved ? pathToFileURL(saved.path).href : `attachment:///${encodeURIComponent(d.name)}`;
        out.blocks.push({ type: 'resource', resource: { uri, mimeType: 'text/plain', text: d.text } });
      }
      out.attachments.push({ kind: 'text', blob: saved?.name, name: d.name });
    } else {
      const image = await readImageFile(d.uri);
      if (image) {
        // The agent said it takes no images — a resource_link is not a fallback, it reads the file as pixels anyway
        if (noImages) { out.problems.push(t('host.imageUnsupported', { name: d.name })); continue; }
        const saved = await stage(extOfMime(image.mimeType), image.bytes, d.name);
        out.blocks.push({ type: 'image', mimeType: image.mimeType, data: image.bytes.toString('base64') });
        out.attachments.push({ kind: 'image', blob: saved?.name, mimeType: image.mimeType, name: d.name });
      } else {
        out.blocks.push({ type: 'resource_link', uri: d.uri, name: d.name || basename(d.uri) });
        out.attachments.push({ kind: 'file', uri: d.uri, name: d.name });
      }
    }
  }
  return out;
}

// The inverse of preparePrompt, for sending a persisted user turn again: blobs are read back into image / text drafts, files stay links.
// An image that came from a file on disk was persisted as an image blob, so it goes out as pixels again rather than being re-read from the original path.
// Attachments whose blob never made it to disk cannot come back and are left out
export async function restoreDrafts(sessionId: string, attachments: Attachment[], blobs: BlobStore): Promise<Draft[]> {
  const drafts = await Promise.all(attachments.map(async (a): Promise<Draft | undefined> => {
    if (a.kind === 'file') return { kind: 'file', uri: a.uri, name: a.name };
    if (!a.blob) return undefined;
    const bytes = Buffer.from(await blobs.readBlob(sessionId, a.blob));
    return a.kind === 'image'
      ? { kind: 'image', mimeType: a.mimeType, data: bytes.toString('base64'), name: a.name }
      : { kind: 'text', name: a.name, text: bytes.toString('utf8') };
  }));
  return drafts.filter((d): d is Draft => d !== undefined);
}

// A file:// URI that names an image the models accept, small enough to inline; anything else (non-image, unreadable, oversized, remote) yields undefined.
// Size is checked before reading so an oversized file never lands in memory
async function readImageFile(uri: string): Promise<{ mimeType: string; bytes: Buffer } | undefined> {
  const mimeType = imageMimeOf(uri);
  if (!mimeType || !uri.startsWith('file:')) return undefined;
  try {
    const path = fileURLToPath(uri);
    const { size } = await stat(path);
    return size <= MAX_IMAGE_BYTES ? { mimeType, bytes: await readFile(path) } : undefined;
  } catch {
    return undefined;
  }
}

// One-line description of what a prompt carried (drafts before staging or attachments after), for the title of a session opened with attachments only
export function describeDrafts(drafts: (Draft | Attachment)[]): string {
  const images = drafts.filter(d => d.kind === 'image').length;
  const files = drafts.filter(d => d.kind !== 'image').map(d => d.name);
  return [images ? t('host.images', { n: images }) : '', ...files].filter(Boolean).join(t('common.listSep'));
}

File: rust/crates/acpira-host/src/acp/attachments.rs
//! Prompt staging (mirror of src/host/acp/attachments.ts): composer text + drafts → the wire prompt and the transcript attachments

use base64::Engine;
use serde_json::{Value, json};

use acpira_shared::attachments::{MAX_IMAGE_BYTES, MAX_TEXT_BYTES, base64_bytes, ext_of_mime, image_mime_of};
use acpira_shared::transcript::{Attachment, Draft};

use super::agent_registry::AgentDef;
use super::normalize::file_url_to_path;
use crate::i18n::{t, tp};
use crate::json::basename;
use crate::store::transcript_store::TranscriptStore;

const B64: base64::engine::GeneralPurpose = base64::engine::general_purpose::STANDARD;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct PreparedPrompt {
  pub blocks: Vec<Value>,
  pub attachments: Vec<Attachment>,
  pub problems: Vec<String>,
}

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct PromptCaps {
  pub embedded_context: bool,
  pub image: bool,
  pub images_regardless: bool,
}

/// The advertised set, defaulting to permitted when the agent said nothing, overridden by the registry
pub fn prompt_caps_of(init: Option<&Value>, def: &AgentDef) -> PromptCaps {
  let p = init.and_then(|i| i.get("agentCapabilities")).and_then(|c| c.get("promptCapabilities"));
  let flag = |k: &str| p.and_then(|p| p.get(k)).and_then(Value::as_bool).unwrap_or(true);
  PromptCaps { embedded_context: flag("embeddedContext"), image: flag("image"), images_regardless: def.images_regardless }
}

pub async fn prepare_prompt(
  session_id: &str,
  text: &str,
  drafts: &[Draft],
  blobs: &TranscriptStore,
  caps: Option<PromptCaps>,
) -> PreparedPrompt {
  let mut out =
    PreparedPrompt { blocks: if text.is_empty() { vec![] } else { vec![json!({ "type": "text", "text": text })] }, ..Default::default() };
  let no_images = caps.is_some_and(|c| !c.image && !c.images_regardless);
  for d in drafts {
    match d {
      Draft::Image { mime_type, data, name } => {
        let label = name.clone().unwrap_or_else(|| t("common.image"));
        if no_images {
          out.problems.push(tp("host.imageUnsupported", &[("name", &label)]));
          continue;
        }
        if base64_bytes(data) > MAX_IMAGE_BYTES {
          out.problems.push(tp("host.imageTooBig", &[("name", &label), ("mb", &(MAX_IMAGE_BYTES >> 20).to_string())]));
          continue;
        }
        let bytes = B64.decode(data.as_bytes()).unwrap_or_default();
        let saved = stage(&mut out, blobs, session_id, ext_of_mime(mime_type), &bytes, &label).await;
        out.blocks.push(json!({ "type": "image", "mimeType": mime_type, "data": data }));
        out.attachments.push(Attachment::Image { blob: saved.map(|s| s.0), mime_type: mime_type.clone(), name: name.clone() });
      }
      Draft::Text { name, text } => {
        if text.len() > MAX_TEXT_BYTES {
          out.problems.push(tp("attach.tooBigText", &[("name", name), ("kb", &(MAX_TEXT_BYTES >> 10).to_string())]));
          continue;
        }
        let saved = stage(&mut out, blobs, session_id, ".txt", text.as_bytes(), name).await;
        if caps.is_some_and(|c| !c.embedded_context) {
          out.blocks.push(json!({ "type": "text", "text": format!("[Attachment: {name}]\n{text}\n[End of attachment: {name}]") }));
        } else {
          let uri = match &saved {
            Some((_, path)) => path_to_file_url(path),
            None => format!("attachment:///{}", encode_uri_component(name)),
          };
          out.blocks.push(json!({ "type": "resource", "resource": { "uri": uri, "mimeType": "text/plain", "text": text } }));
        }
        out.attachments.push(Attachment::Text { blob: saved.map(|s| s.0), name: name.clone() });
      }
      Draft::File { uri, name } => {
        if let Some((mime, bytes)) = read_image_file(uri).await {
          if no_images {
            out.problems.push(tp("host.imageUnsupported", &[("name", name)]));
            continue;
          }
          let saved = stage(&mut out, blobs, session_id, ext_of_mime(mime), &bytes, name).await;
          out.blocks.push(json!({ "type": "image", "mimeType": mime, "data": B64.encode(&bytes) }));
          out.attachments.push(Attachment::Image { blob: saved.map(|s| s.0), mime_type: mime.to_owned(), name: Some(name.clone()) });
        } else {
          let link_name = if name.is_empty() { basename(uri) } else { name.clone() };
          out.blocks.push(json!({ "type": "resource_link", "uri": uri, "name": link_name }));
          out.attachments.push(Attachment::File { uri: uri.clone(), name: name.clone() });
        }
      }
    }
  }
  out
}

async fn stage(
  out: &mut PreparedPrompt,
  blobs: &TranscriptStore,
  session_id: &str,
  ext: &str,
  bytes: &[u8],
  label: &str,
) -> Option<(String, String)> {
  match blobs.save_blob(session_id, ext, bytes).await {
    Ok((name, path)) => Some((name, path.to_string_lossy().into_owned())),
    Err(e) => {
      out.problems.push(tp("host.attachStageFailed", &[("label", label), ("error", &e.to_string())]));
      None
    }
  }
}

/// The inverse of prepare_prompt, for sending a persisted user turn again
pub async fn restore_drafts(session_id: &str, attachments: &[Attachment], blobs: &TranscriptStore) -> anyhow::Result<Vec<Draft>> {
  let mut out = vec![];
  for a in attachments {
    match a {
      Attachment::File { uri, name } => out.push(Draft::File { uri: uri.clone(), name: name.clone() }),
      Attachment::Image { blob: Some(b), mime_type, name } => {
        let bytes = blobs.read_blob(session_id, b).await?;
        out.push(Draft::Image { mime_type: mime_type.clone(), data: B64.encode(bytes), name: name.clone() });
      }
      Attachment::Text { blob: Some(b), name } => {
        let bytes = blobs.read_blob(session_id, b).await?;
        out.push(Draft::Text { name: name.clone(), text: String::from_utf8_lossy(&bytes).into_owned() });
      }
      _ => {}
    }
  }
  Ok(out)
}

async fn read_image_file(uri: &str) -> Option<(&'static str, Vec<u8>)> {
  let mime = image_mime_of(uri)?;
  if !uri.starts_with("file:") {
    return None;
  }
  let path = file_url_to_path(uri)?;
  let meta = tokio::fs::metadata(&path).await.ok()?;
  if meta.len() as usize > MAX_IMAGE_BYTES {
    return None;
  }
  Some((mime, tokio::fs::read(&path).await.ok()?))
}

/// One-line description of what a prompt carried
pub fn describe_drafts(items: &[Attachment]) -> String {
  let images = items.iter().filter(|d| matches!(d, Attachment::Image { .. })).count();
  let mut parts: Vec<String> = vec![];
  if images > 0 {
    parts.push(tp("host.images", &[("n", &images.to_string())]));
  }
  for d in items {
    match d {
      Attachment::Text { name, .. } | Attachment::File { name, .. } if !name.is_empty() => parts.push(name.clone()),
      _ => {}
    }
  }
  parts.join(&t("common.listSep"))
}

pub fn describe_draft_list(items: &[Draft]) -> String {
  let images = items.iter().filter(|d| matches!(d, Draft::Image { .. })).count();
  let mut parts: Vec<String> = vec![];
  if images > 0 {
    parts.push(tp("host.images", &[("n", &images.to_string())]));
  }
  for d in items {
    match d {
      Draft::Text { name, .. } | Draft::File { name, .. } if !name.is_empty() => parts.push(name.clone()),
      _ => {}
    }
  }
  parts.join(&t("common.listSep"))
}

/// url.pathToFileURL(p).href
pub fn path_to_file_url(p: &str) -> String {
  let path = if cfg!(windows) { format!("/{}", p.replace('\\', "/")) } else { p.to_owned() };
  let mut out = String::from("file://");
  for b in path.bytes() {
    let keep = b.is_ascii_alphanumeric() || b"/-._~!$&'()*+,;=:@".contains(&b);
    if keep {
      out.push(b as char);
    } else {
      out.push_str(&format!("%{b:02X}"));
    }
  }
  out
}

/// encodeURIComponent
pub fn encode_uri_component(s: &str) -> String {
  let mut out = String::new();
  for b in s.bytes() {
    if b.is_ascii_alphanumeric() || b"-_.!~*'()".contains(&b) {
      out.push(b as char);
    } else {
      out.push_str(&format!("%{b:02X}"));
    }
  }
  out
}
